### DISCUSSION TITLE
Specify a url when return a JSON response

### DISCUSSION THREAD

**Discussion by brightpixels** [remix-run/remix#6787] [Q&A]
Hi,
In one of our pages, when we have an error (e.g. user didn't choose something from a drop down), we return a [JSON response ](https://remix.run/docs/en/main/utils/json) with a status 400, which reloads the pages with some errors in the UI, however, the url is stripped of the url parameters that were present initially.  In our code logic, we need the url parameters because we use it as a signal to re-fetch data needed to populate the drop down again. It looks like JSON response you can specify a url (a bit like redirect response) but it doesn't have an efffect:

```
return json(responseData, { status: 400, headers: {
    url: "?#step-2"
  } });
```

Any ideas how to return a JSON response but with a url/parameters (a bit like redirect response) since we need to return the user to the same page, with some data, status 400, and with the same url before the error

**Comment by kiliman** [comment]
I'm trying to follow what you're saying, but it would be helpful if you could post a simple repo showing what you're trying to do.

**Comment by sergiodxa** [comment]
From what I understood, what you need to do is either POST directly to the final URL (including any search param) or do a redirect, save the data in the session, then read it in the loader if send a 400 response from the loader.

```ts
export async function action({ request }: DataFunctionArgs) {
  let session = await sessionStorage.getSession(
    request.headers.get("cookie")
  )

  session.flash("error", "error message or other data, it can be JSON")

  let headers = new Headers()
  headers.append("set-cookie", await sessionStorage.commitSession(session)

  return redirect("/some/route?param=value", { headers })
}

export async function loader({ request }: DataFunctionArgs) {
  let session = await sessionStorage.getSession(
    request.headers.get("cookie")
  )

  let error = session.get("error")

  let data = await getDataForThisRoute()
  
  if (error) return json({ data, error }, 400) // return 400 with error
  return json({ data })
}
```
